src/Lib.P2P/AutoDialer.cs

```diff
@@ -0,0 +1,193 @@
+        /// </param>
+        private void Dispose(bool disposing)
+        {
+            if (!disposing)
+            {
+                return;
+            }
+            
+            _swarmService.PeerDiscovered -= OnPeerDiscovered;
+            _swarmService.PeerDisconnected -= OnPeerDisconnected;
+        }
+
+        /// <summary>
+        ///   Performs application-defined tasks associated with freeing, 
+        ///   releasing, or resetting unmanaged resources.
+        /// </summary>
+        public void Dispose() { Dispose(true); }
+
+        /// <summary>
+        ///   The low water mark for peer connections.
+        /// </summary>
+        /// <value>
+        ///   Defaults to <see cref="DefaultMinConnections"/>.
+        /// </value>
+        /// <remarks>
+        ///   Setting this to zero will basically disable the auto dial features.
+        /// </remarks>
+        public int MinConnections { get; set; } = DefaultMinConnections;
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        /// <summary>
+        ///   Called when the swarm has a new peer.
+        /// </summary>
+        /// <param name="sender">
+        ///   The swarm of peers.
+        /// </param>
+        /// <param name="peer">
+        ///   The peer that was discovered.
+        /// </param>
+        /// <remarks>
+        ///   If the <see cref="MinConnections"/> is not reached, then the
+        ///   <paramref name="peer"/> is dialed.
+        /// </remarks>
+        private async void OnPeerDiscovered(object sender, Peer peer)
+#pragma warning restore VSTHRD100 // Avoid async void methods
+        {
+            var n = _swarmService.Manager.Connections.Count() + _pendingConnects;
+            
+            if (!_swarmService.IsRunning || n >= MinConnections)
+            {
+                return;
+            }
+            
+            Interlocked.Increment(ref _pendingConnects);
+            Log.Debug($"Dialing new {peer}");
+            try
+            {
+                await _swarmService.ConnectAsync(peer).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                Log.Warn($"Failed to dial {peer}");
+            }
+            finally
+            {
+                Interlocked.Decrement(ref _pendingConnects);
+            }
+        }
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        /// <summary>
+        ///   Called when the swarm has lost a connection to a peer.
+        /// </summary>
+        /// <param name="sender">
+        ///   The swarm of peers.
+        /// </param>
+        /// <param name="disconnectedPeer">
+        ///   The peer that was disconnected.
+        /// </param>
+        /// <remarks>
+        ///   If the <see cref="MinConnections"/> is not reached, then another
+        ///   peer is dialed.
+        /// </remarks>
+        private async void OnPeerDisconnected(object sender, Peer disconnectedPeer)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this unused method parameter 'sender'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)